File: 140-word-break-ii/word-break-ii.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        word_set = set(wordDict)
        out = []

        def dfs(start: int, path: List[str]):
            # base case
            if start == len(s):
                path_str = " ".join(path)
                out.append(path_str)
                return

            # recursive case
            for i in range(start, len(s)):
                sub_str = s[start:i + 1]
                if sub_str in word_set:
                    path.append(sub_str)
                    dfs(i + 1, path)
                    path.pop()

        dfs(0, [])
        return out

File: 140-word-break-ii/word-break-ii.go
func wordBreak(s string, wordDict []string) []string {
    set := map[string]struct{}{}
    for i := 0; i < len(wordDict); i++ {
        set[wordDict[i]] = struct{}{}
    }
    out := []string{}
    var dfs func(start int, path []string)
    dfs = func(start int, path []string) {
        // base
        if start == len(s) {
            pathStr := strings.Join(path, " ")
            out = append(out, pathStr)
            return
        }
         
        // logic
        for i := start ; i < len(s); i++ {
            subStr := s[start:i+1]
            if _, ok := set[subStr]; ok {
                path = append(path, subStr)
                dfs(i+1, path)
                path = path[:len(path)-1]
            }
        }
    }
    dfs(0, nil)
    return out
}